controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -317,7 +378,50 @@ public String getScopeName() {
     public CompletableFuture<Void> scaleNewSegmentsCreated(final List<Integer> sealedSegments,
                                                            final List<Integer> newSegments,
                                                            final long scaleTimestamp) {
-        return verifyLegalState(FutureHelpers.toVoid(addPartialHistoryRecord(sealedSegments, newSegments)));
+        return verifyState(() -> FutureHelpers.toVoid(addPartialHistoryRecord(sealedSegments, newSegments)),
+                Lists.newArrayList(State.SCALING));
+    }
+
+    /**
+     * If scale is ongoing, try to delete the epoch node.
+     *
+     * @param epoch epoch
+     * @return true if we are able to delete the epoch, false otherwise.
+     */
+    @Override
+    public CompletableFuture<Boolean> scaleTryDeleteEpoch(final int epoch) {
+        return getHistoryTableFromStore()
+                .thenCompose(historyTable -> getSegmentTableFromStore().thenApply(segmentTable -> new ImmutablePair<>(historyTable, segmentTable)))
+                .thenCompose(pair -> {
+                    Data<T> segmentTable = pair.getRight();
+                    Data<T> historyTable = pair.getLeft();
+                    CompletableFuture<Boolean> result = new CompletableFuture<>();
+
+                    if (TableHelper.isScaleOngoing(historyTable.getData(), segmentTable.getData())) {
+                        deleteEpochNode(epoch)
+                                .whenComplete((r, e) -> {
+                                    if (e != null) {
+                                        Throwable ex = ExceptionHelpers.getRealException(e);
+                                        if (ex instanceof DataExistsException) {
+                                            // cant delete as there are transactions still running under epoch node
+                                            result.complete(false);
+                                        } else {
+                                            result.completeExceptionally(ex);
```

Comment:
If the epoch node is already deleted, we should try to complete remaining steps and not quit with an exception. This is because, some other process could have failed immediately after deleting the epoch node but before executing the subsequent steps.